world/shared_task_manager.cpp

```diff
@@ -0,0 +1,1820 @@
+				s->GetDbSharedTask().id,
+				character_name
+			);
+
+			m.is_leader = true;
+		}
+	}
+
+	if (found_new_leader) {
+		s->SetMembers(members);
+		SaveMembers(s, members);
+		SendSharedTaskMemberListToAllMembers(s);
+		SendMembersMessageID(
+			s, Chat::Yellow, SharedTaskMessage::PLAYER_NOW_LEADER,
+			{new_leader.character_name, s->GetTaskData().title}
+		);
+
+		for (const auto &dz_id : s->dynamic_zone_ids) {
+			auto dz = DynamicZone::FindDynamicZoneByID(dz_id);
+			if (dz) {
+				dz->SetNewLeader(static_cast<uint32_t>(new_leader.character_id));
+			}
+		}
+	}
+}
+
+void SharedTaskManager::SaveMembers(SharedTask *s, std::vector<SharedTaskMember> members)
+{
+	std::vector<SharedTaskMembersRepository::SharedTaskMembers> dm = {};
+	dm.reserve(members.size());
+	for (auto &m: members) {
+		auto e = SharedTaskMembersRepository::NewEntity();
+
+		e.character_id   = m.character_id;
+		e.is_leader      = (m.is_leader ? 1 : 0);
+		e.shared_task_id = s->GetDbSharedTask().id;
+
+		dm.emplace_back(e);
+	}
+
+	SharedTaskMembersRepository::DeleteWhere(*m_database, fmt::format("shared_task_id = {}", s->GetDbSharedTask().id));
+	SharedTaskMembersRepository::InsertMany(*m_database, dm);
+}
+
+void SharedTaskManager::InvitePlayerByPlayerName(SharedTask *s, const std::string &player_name)
+{
+	auto character = CharacterDataRepository::GetWhere(
+		*m_database,
+		fmt::format("`name` = '{}' LIMIT 1", EscapeString(player_name))
+	);
+
+	auto character_id = !character.empty() ? character.front().id : 0;
+
+	// we call validation even for an invalid player so error messages occur
+	if (CanAddPlayer(s, character_id, player_name, false)) {
+		// send dialogue window
+		SendSharedTaskInvitePacket(s, character_id);
+
+		// keep track of active invitations at world
+		QueueActiveInvitation(s->GetDbSharedTask().id, character_id);
+	}
+}
+
+void SharedTaskManager::SendSharedTaskInvitePacket(SharedTask *s, int64 invited_character_id)
+{
+	auto leader = s->GetLeader();
+
+	// found leader
+	if (leader.character_id > 0) {
+
+		// init packet
+		auto p = std::make_unique<ServerPacket>(
+			ServerOP_SharedTaskInvitePlayer,
+			sizeof(ServerSharedTaskInvitePlayer_Struct)
+		);
+
+		// fill
+		auto d = reinterpret_cast<ServerSharedTaskInvitePlayer_Struct *>(p->pBuffer);
+		d->requested_character_id = invited_character_id;
+		d->invite_shared_task_id  = s->GetDbSharedTask().id;
+		strn0cpy(d->inviter_name, leader.character_name.c_str(), sizeof(d->inviter_name));
+		strn0cpy(d->task_name, s->GetTaskData().title.c_str(), sizeof(d->task_name));
+
+		// get requested character zone server
+		ClientListEntry *cle = client_list.FindCLEByCharacterID(invited_character_id);
+		if (cle && cle->Server()) {
+			SendLeaderMessageID(s, Chat::Yellow, SharedTaskMessage::SEND_INVITE_TO, {cle->name()});
+			cle->Server()->SendPacket(p.get());
+		}
+	}
+}
+
+void SharedTaskManager::AddPlayerByCharacterIdAndName(
+	SharedTask *s,
+	int64 character_id,
```

Comment:
Another one, we can probably punt this and clean stuff like this up in another branch though